### PR TITLE
Enforce that facet "filter-operator" is a valid value

### LINKED ISSUE #3027: Enforce that facet "filter-operator" is a valid value
_Originally posted by @KevinVdV in https://github.com/DSpace/dspace-angular/issues/915#issuecomment-713540955_

In https://github.com/DSpace/dspace-angular/issues/915, an invalid request is being sent from the Angular UI to the REST API because the "operator" appears to be missing.

On the REST side we should be enforcing the operator to have a valid value (for the request in https://github.com/DSpace/dspace-angular/issues/915 the operator would be "kb" which is invalid). What I would suggest is to return a bad request if the operator is missing or invalid.

I think the work for this is half a day (the change is small but we would also require tests)

### PR DESCRIPTION
**Pull request by KevinVdV**
## References
* Fixes #3027

## Description
If no valid operator is passed in a filter query the discovery endpoint will now result in a bad request.

In my original issue above I noted that this shouldn't be checked for dates, but I'm afraid dates will also have to follow these rules (contains is the default), because the java code has no way to tell if the field is a date or not.

## Instructions for Reviewers
Fire of a query similar to: {dspace.url}/api/discover/search/objects?sort=score,DESC&page=0&size=10&f.author=pujari,%20kb

This should result in a bad request since the operator would be "kb". Next use:

 {dspace.url}/api/discover/search/objects?sort=score,DESC&page=0&size=10&f.author=pujari,%20kb,contains

This should work as "contains" is a valid operator.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [X] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [X] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [X] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [X] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [X] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [X] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me. I also tested it and it works as described.  Thanks @KevinVdV !

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/parameter/resolver/SearchFilterResolver.java [hunk 1]
```diff
@@ -61,6 +65,19 @@ public Object resolveArgument(final MethodParameter parameter, final ModelAndVie
...
 
+    private void checkIfValidOperator(String filterOperator) {
+        if (StringUtils.isNotBlank(filterOperator)) {
+            if (!ALLOWED_SEARCH_OPERATORS.contains(filterOperator.trim())) {
+                throw new DSpaceBadRequestException(
+                    "The operator can't be \"" + filterOperator + "\", must be the of one of: " +
+                    String.join(", ", ALLOWED_SEARCH_OPERATORS));
+            }
+        } else {
+            throw new DSpaceBadRequestException(
```
Comment (abollini) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/parameter/resolver/SearchFilterResolver.java:76]: I suggest to return an UnprocessableEntityException here

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 2]
```diff
@@ -2895,6 +3197,108 @@ public void discoverSearchObjectsWithQueryOperatorNotAuthority() throws Exceptio
...
+        //** WHEN **
+        // An anonymous user browses this endpoint to find the the objects in the system
+        // With the given search filter where there is a non-valid filter operator given (must be of form
+        // <:filter-value>,<:filter-operator> where the filter operator is one of: “contains”, “notcontains”, "equals"
+        // “notequals”, “authority”, “notauthority”, "query”); see enum RestSearchOperator
+        getClient().perform(get("/api/discover/search/objects")
+            .param("f.title", "test,operator"))
+                   //** THEN **
+                   //Will result in 404 status because of non-valid filter operator
+                   .andExpect(status().isBadRequest());
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:3299]: this should be a 422 (please fix also the comment)

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 3]
```diff
@@ -2895,6 +3197,108 @@ public void discoverSearchObjectsWithQueryOperatorNotAuthority() throws Exceptio
...
+    @Test
+    public void discoverSearchObjectsWithMissingQueryOperator() throws Exception {
+        //** WHEN **
+        // An anonymous user browses this endpoint to find the the objects in the system
+        // With the given search filter where there is the filter operator missing in the value (must be of form
+        // <:filter-value>,<:filter-operator>)
+        getClient().perform(get("/api/discover/search/objects")
+            .param("f.title", "test"))
+                   //** THEN **
+                   //Will result in 404 status because of missing filter operator
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:3284]: please update the comment to reflect 400

**Review by abollini — CHANGES_REQUESTED** [review]
thanks @KevinVdV for this PR I agree with the proposed changes at the implementation level but I think the error code here should be 422. This because if the parameter is here and it is a string the error is more "conceptually" so according to what we have also recently discussed and summarized here https://github.com/DSpace/DSpace/issues/3093#issuecomment-745347787 422 looks more appropriate to me.

**Comment by abollini** [comment]
I noted that the rest contract doesn't list return code yet https://github.com/DSpace/Rest7Contract/blob/main/search-endpoint.md#matching-dspace-objects-search-results so it would be useful to also make a PR to add these details now that we are finalizing them

**Comment by KevinVdV** [comment]
@abollini We fixed your test feedback & I created a PR for the rest contract here: https://github.com/DSpace/Rest7Contract/pull/154

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/parameter/resolver/SearchFilterResolver.java (hunk 1 above)
Comment (abollini) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/parameter/resolver/SearchFilterResolver.java:77]: this comment has been not yet processed

File: dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java [hunk 4]
```diff
@@ -2895,6 +3197,108 @@ public void discoverSearchObjectsWithQueryOperatorNotAuthority() throws Exceptio
...
+    public void discoverSearchObjectsWithMissingQueryOperator() throws Exception {
+        //** WHEN **
+        // An anonymous user browses this endpoint to find the the objects in the system
+        // With the given search filter where there is the filter operator missing in the value (must be of form
+        // <:filter-value>,<:filter-operator>)
+        getClient().perform(get("/api/discover/search/objects")
+            .param("f.title", "test"))
+                   //** THEN **
+                   //Will result in 400 status because of missing filter operator
+                   .andExpect(status().isBadRequest());
```
Comment (abollini) [inline dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:3285]: this should be 422 as well as agreed in the rest contract

**Review by abollini — CHANGES_REQUESTED** [review]
hi @KevinVdV  @MarieVerdonck  there is still a small inconsistency compared to the rest contract. When the request parameter is here without a filter operator 422 should be returned, see inline unresolved comments

**Comment by KevinVdV** [comment]
@abollini We fixed your latest feedback.

**Review by abollini — APPROVED** [review]

**Comment by abollini** [comment]
thanks @MarieVerdonck @KevinVdV it looks perfect now

**Pull request merged**